HeterogeneousCore/CUDAUtilities/src/CUDAEventCache.cc

```diff
@@ -0,0 +1,74 @@
+#include "FWCore/Utilities/interface/thread_safety_macros.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/CUDAEventCache.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/cudaCheck.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/currentDevice.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/ScopedSetDevice.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/cudaDeviceCount.h"
+
+namespace cudautils {
+  void CUDAEventCache::Deleter::operator()(cudaEvent_t event) const {
+    if (device_ != -1) {
+      ScopedSetDevice deviceGuard{device_};
+      cudaCheck(cudaEventDestroy(event));
+    }
+  }
+
+  // CUDAEventCache should be constructed by the first call to
+  // getCUDAEventCache() only if we have CUDA devices present
+  CUDAEventCache::CUDAEventCache() : cache_(cudautils::cudaDeviceCount()) {}
+
+  SharedEventPtr CUDAEventCache::getCUDAEvent() {
+    const auto dev = cudautils::currentDevice();
+    auto event = makeOrGet(dev);
+    auto ret = cudaEventQuery(event.get());
+    // event is occurred, return immediately
+    if (ret == cudaSuccess) {
+      return event;
+    }
+    // return code is something else than "recorded", throw exception
+    if (ret != cudaErrorNotReady) {
+      cudaCheck(ret);
+    }
+
+    // Got recorded, but not yet occurred event. Try until we get an
+    // occurred event. Need to keep all recorded events until an
+    // occurred event is found in order to avoid ping-pong with a
+    // recorded event.
+    std::vector<SharedEventPtr> ptrs{std::move(event)};
+    do {
```

Comment:
As soon as the `ReusableObjectHolder` makes a new object, the newly created `cudaEvent_t` will get returned. But theoretically it could grow unbounded.

Reply:
Why would `makeOrGet()` return an event with pending work on it (i.e. not yet "occurred") ?
I thought I understood that it would either create a new event, or return an existing one that has been "recycled" after it has no work left pending ?

Reply:
Currently it is technically possible to return a "recorded" / "work incomplete" event:
```cpp
void func(cudaStream_t stream) {
   some_long_running_kernel<<<...,stream>>>(...);
   auto event = cudautils::getCUDAEventCache().getCUDAEvent();
   cudaCheck(cudaEventRecord(event.get());
  // event destructed and returned to cache in "recorded" / "work incomplete" state
}
```
I believe such use would be mostly a mistake (why bother with an `cudaEvent_t` if one doesn't synchronize?), but if done it has high potential of messing up things elsewhere.

Thinking out loud some possibilities

1. could add a call to `cudaEventSynchronize()` to the deleter of the `shared_ptr` returned by `edm::ReusableObjectHolder`
   * would require some work on `ReusableObjectHolder`
   * would introduce a possible blocking synchronization on the destructor of the last `shared_ptr` instance
       * on the other hand could be acceptable if we decide such a use case to be a mistake
    * in principle would make the destructor of the `shared_ptr` throwing (should check the return value of `cudaEventSynchronize()`)
2. same as 1, but instead of `cudaEventSynchronize()` do `assert(cudaEventQuery() == cudaSuccess)`
   * mostly the same caveats as in 1

Reply:
I was wrong, there is actually a valid use-case for returning an incomplete event to the cache: using `cudaStreamWaitEvent()` for the synchronization
```cpp
auto func1(cudaStream_t stream) {
   some_long_running_kernel<<<..., stream>>>(...);
   auto event = cudautils::getCUDAEventCache().getCUDAEvent();
  cudaCheck(cudaEventRecord(event.get());
  return event;
}

void func2(cudaStreamt_t stream1, cudaStream_t stream2) {
   auto event = func1(stream1);
   cudaCheck(cudaStreamWaitEvent(stream2, event.get()));
   another_kernel<<<..., stream2>>>(...);
}
```
Such a code path could actually occur if we ever get to delete `CUDAProduct<T>`'s early.

An alternative for collecting incomplete events here could be to `cudaEventDestroy()` them in the `shared_ptr` deleter when they are incomplete instead of returning to the cache. According to the [docs](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__EVENT.html#group__CUDART__EVENT_1g2cb6baa0830a1cd0bd957bfd8705045b) 
> An event may be destroyed before it is complete (i.e., while cudaEventQuery() would return cudaErrorNotReady). In this case, the call does not block on completion of the event, and any associated resources will automatically be released asynchronously at completion.

But I think I'll rather keep the current solution than go with `cudaEventDestroy()`.

Reply:
@makortel in this example, `func1` does return an incomplete event - but IIUC it is not returned to the cache until the end of `func2`, when it is no longer incomplete - is it ?

Reply:
The `event` may be incomplete at the end of `func2` (if `some_long_running_kernel()` is still being run on the GPU at that point). AFAIK `cudaStreamWaitEvent()` only makes the `stream2` to wait for the `event` to become complete, but does not synchronize the host thread.

Reply:
Right, I see.